packages/frontend/editor-ui/src/components/canvas/experimental/components/ExperimentalEmbeddedNdvMapper.vue

```diff
@@ -1,57 +1,105 @@
 <script setup lang="ts">
 import InputPanel from '@/components/InputPanel.vue';
-import { CanvasKey } from '@/constants';
 import type { INodeUi } from '@/Interface';
 import { useNDVStore } from '@/stores/ndv.store';
-import { useCanvasStore } from '@/stores/canvas.store';
-import { onBeforeUnmount, watch } from 'vue';
+import { onBeforeUnmount, ref, watch } from 'vue';
 import type { Workflow } from 'n8n-workflow';
-import { computed, inject, useTemplateRef } from 'vue';
+import { computed, useTemplateRef } from 'vue';
 import { N8nPopoverReka } from '@n8n/design-system';
 import { useStyles } from '@/composables/useStyles';
+import {
+	onClickOutside,
+	useElementHover,
+	type UseElementHoverOptions,
+	useEventListener,
+} from '@vueuse/core';
+import { useExperimentalNdvStore } from '@/components/canvas/experimental/experimentalNdv.store';
 
-const { node, inputNodeName, visible, virtualRef } = defineProps<{
+type MapperState = { isOpen: true; closeOnMouseLeave: boolean } | { isOpen: false };
+
+const hoverOptions: UseElementHoverOptions = {
+	delayLeave: 200, // should be a positive value, otherwise user cannot click on the mapper
+};
+
+const {
+	node,
+	inputNodeName,
+	reference,
+	visibleOnHover = false,
+} = defineProps<{
 	workflow: Workflow;
 	node: INodeUi;
 	inputNodeName: string;
-	visible: boolean;
-	virtualRef: HTMLElement | undefined;
+	visibleOnHover?: boolean;
+	reference: HTMLElement;
 }>();
 
+const state = ref<MapperState>({ isOpen: false });
 const contentRef = useTemplateRef('content');
 const ndvStore = useNDVStore();
-const canvas = inject(CanvasKey, undefined);
-const isVisible = computed(() => visible && !canvas?.isPaneMoving.value);
-const canvasStore = useCanvasStore();
-const contentElRef = computed(() => contentRef.value?.$el ?? null);
+const experimentalNdvStore = useExperimentalNdvStore();
+const contentElRef = computed<HTMLElement | null>(() => contentRef.value?.$el ?? null);
 const { APP_Z_INDEXES } = useStyles();
+const isReferenceHovered = useElementHover(visibleOnHover ? reference : null, hoverOptions);
+const isMapperHovered = useElementHover(visibleOnHover ? contentElRef : null, hoverOptions);
+const isHovered = computed(() => isReferenceHovered.value || isMapperHovered.value);
+
+function handleFocusIn() {
+	if (!experimentalNdvStore.isMapperOpen) {
+		state.value = { isOpen: true, closeOnMouseLeave: false };
+	}
+}
+
+function handleReferenceFocusOut(event: FocusEvent | MouseEvent) {
+	if (
+		!(event.target instanceof Node && reference?.contains(event.target)) &&
+		!(event.target instanceof Node && contentElRef.value?.contains(event.target)) &&
+		!(event.relatedTarget instanceof Node && contentElRef.value?.contains(event.relatedTarget))
```

Comment:
nit: Could we extract this into something like `isContainedWithin` or similar to make it easier to read?